ecdsa_go1.14.go

```diff
@@ -0,0 +1,77 @@
+//+build !go1.15
+
+package jwt
+
+import (
+	"crypto/ecdsa"
+	"crypto/rand"
+	"math/big"
+	"math/bits"
+)
+
+// Implements the Sign method from SigningMethod
+// For this signing method, key must be an ecdsa.PrivateKey struct
+func (m *SigningMethodECDSA) Sign(signingString string, key interface{}) (string, error) {
+	// Get the key
+	var ecdsaKey *ecdsa.PrivateKey
+	switch k := key.(type) {
+	case *ecdsa.PrivateKey:
+		ecdsaKey = k
+	default:
+		return "", ErrInvalidKeyType
+	}
+
+	// Create the hasher
+	if !m.Hash.Available() {
+		return "", ErrHashUnavailable
+	}
+
+	hasher := m.Hash.New()
+	hasher.Write([]byte(signingString))
+
+	// Sign the string and return r, s
+	if r, s, err := ecdsa.Sign(rand.Reader, ecdsaKey, hasher.Sum(nil)); err == nil {
+		curveBits := ecdsaKey.Curve.Params().BitSize
+
+		if m.CurveBits != curveBits {
+			return "", ErrInvalidKey
+		}
+
+		keyBytes := curveBits / 8
+		if curveBits%8 > 0 {
+			keyBytes += 1
+		}
+
+		// We serialize the outputs (r and s) into big-endian byte arrays
+		// padded with zeros on the left to make sure the sizes work out.
+		// Output must be 2*keyBytes long.
+		out := make([]byte, 2*keyBytes)
+		fillBytesInt(r, out[0:keyBytes]) // r is assigned to the first half of output.
+		fillBytesInt(s, out[keyBytes:])  // s is assigned to the second half of output.
+
+		return EncodeSegment(out), nil
+	} else {
+		return "", err
+	}
+}
+
+func fillBytesInt(x *big.Int, buf []byte) []byte {
+	// Clear whole buffer. (This gets optimized into a memclr.)
+	for i := range buf {
+		buf[i] = 0
+	}
+	i := len(buf)
+	for _, d := range x.Bits() {
+		for j := 0; j < (bits.UintSize / 8); j++ {
+			i--
+			if i >= 0 {
+				buf[i] = byte(d)
+			} else if byte(d) != 0 {
+				panic("math/big: buffer too small to fit value")
```

Comment:
For me this looks like a 1:1 copy out of the Go std lib, even including the particular error here and the comment above. Not quite sure, if this ok from a license point of view. Even though it is BSD, it still needs proper attribution and unless you are the original author of that line, this is very fishy in my opinion to just copy/paste that here.

Reply:
Adjusted.

Reply:
I had the exact same thought (was just about to check if this was a copy/paste from the standard library).

I usually reference this ticket by Ian. https://github.com/golang/go/issues/19893#issuecomment-292755517

Reply:
The panic has to be exactly the same otherwise it's not compatible.